api/src/utils/open-payments.ts

```diff
@@ -0,0 +1,468 @@
+      receiveAmount: args.receiveAmount,
+      nonce: clientNonce,
+      paymentId: paymentId,
+      redirectUrl: this.redirectUrl + 'payment-confirmation'
+    })
+
+    return outgoingPaymentGrant
+  }
+
+  async finishPaymentProcess(
+    senderWalletAddress: string,
+    grant: PendingGrant,
+    quote: QuoteResponse,
+    interactRef: string
+  ): Promise<CheckPaymentResult> {
+    const walletAddress = await this.getWalletAddress(senderWalletAddress)
+    const continuation = await this.client!.grant.continue(
+      {
+        url: grant.continue.uri,
+        accessToken: grant.continue.access_token.value
+      },
+      {
+        interact_ref: interactRef
+      }
+    )
+
+    if (!isFinalizedGrant(continuation)) {
+      throw new Error('Expected finalized grant.')
+    }
+
+    const url = new URL(walletAddress.resourceServer).origin
+
+    const outgoingPayment = await this.client!.outgoingPayment.create(
+      {
+        url: url,
+        accessToken: continuation.access_token.value
+      },
+      {
+        walletAddress: walletAddress.id,
+        quoteId: quote.id,
+        metadata: {
+          description: 'Tools Payment'
+        }
+      }
+    ).catch(() => {
+      throw new Error('Could not create outgoing payment.')
+    })
+
+    return await this.checkOutgoingPayment(
+      outgoingPayment.id,
+      continuation.access_token.value,
+      quote.incomingPaymentGrantToken,
+      quote.receiver
+    )
+  }
+
+  private async createIncomingPaymentGrant(urlAuthServer: string) {
+    const nonInteractiveGrant = await this.client!.grant.request(
+      {
+        url: urlAuthServer
+      },
+      {
+        access_token: {
+          access: [
+            {
+              type: 'incoming-payment',
+              actions: ['read', 'create', 'complete']
+            }
+          ]
+        }
+      }
+    )
+
+    if (isPendingGrant(nonInteractiveGrant)) {
+      throw new Error('Expected non-interactive grant')
+    }
+
+    return nonInteractiveGrant
+  }
+
+  private async createIncomingPayment({
+    accessToken,
+    walletAddress,
+    note
+  }: CreateIncomingPaymentParams) {
+    // create incoming payment without amount
+    return await this.client!.incomingPayment.create(
+      {
+        url: walletAddress.resourceServer,
+        accessToken: accessToken
+      },
+      {
+        expiresAt: new Date(Date.now() + 6000 * 60).toISOString(),
+        walletAddress: walletAddress.id,
+        metadata: {
+          description: note
+        }
+      }
+    ).catch(() => {
+      throw new Error('Unable to create incoming payment.')
+    })
+  }
+
+  private async createQuoteGrant({ authServer }: QuoteGrantParams) {
+    return await this.client!.grant.request(
+      {
+        url: authServer
+      },
+      {
+        access_token: {
+          access: [
+            {
+              type: 'quote',
+              actions: ['create', 'read']
+            }
+          ]
+        }
+      }

```

Comment:
Nit: I kinda like these objects formatted more compactly:

```suggestion
      { url: authServer },
      {
        access_token: {
          access: [
            { type: 'quote', actions: ['create', 'read'] }
          ]
        }
      }
```